File: PatchUpdate/BsPatch/app/src/main/java/com/maotou/bspatch/MainActivity.java
package com.maotou.bspatch;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.content.pm.PermissionGroupInfo;
import android.content.pm.PermissionInfo;
import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Environment;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v4.content.PermissionChecker;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.webkit.PermissionRequest;
import android.widget.Toast;

import com.maotou.bspatch.databinding.ActivityMainBinding;

import java.io.File;

public class MainActivity extends AppCompatActivity {

    public static final String SDCARD_PATH = Environment.getExternalStorageDirectory() + File.separator + "Download"  + File.separator;
    public static final String PATCH_FILE = "old-to-new.patch";
    public static final String NEW_APK_FILE = "new.apk";
    private ActivityMainBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        if (ContextCompat.checkSelfPermission(this,Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED){
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE},0);
        }

        
    }

    private void patch(){
        if (ContextCompat.checkSelfPermission(MainActivity.this,Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED) {
            new ApkUpdateTask().executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
        }else{
            Toast.makeText(this,"没有读写权限",Toast.LENGTH_SHORT).show();
        }
    }

    /**
     * 合并增量文件任务
     */
    private class ApkUpdateTask extends AsyncTask<Void, Void, Boolean> {

        @Override
        protected Boolean doInBackground(Void... params) {
            String oldApkPath = ApkUtils.getCurApkPath(MainActivity.this);
            File oldApkFile = new File(oldApkPath);
            File patchFile = new File(getPatchFilePath());
            if(oldApkFile.exists() && patchFile.exists()) {
                Log("正在合并增量文件...");
                String newApkPath = getNewApkFilePath();
                BsPatch.applyPatch(oldApkPath, newApkPath, getPatchFilePath());
//                //检验文件MD5值
//                return Signtils.checkMd5(oldApkFile, MD5);

                Log("增量文件的MD5值为：" + SignUtils.getMd5ByFile(patchFile));
                Log("新文件的MD5值为：" + SignUtils.getMd5ByFile(new File(newApkPath)));

                return true;
            }
            return false;
        }

        @Override
        protected void onPostExecute(Boolean result) {
            super.onPostExecute(result);
            if(result) {
                Log("合并成功，开始安装");
                ApkUtils.installApk(MainActivity.this, getNewApkFilePath());
            } else {
                Log("合并失败");
            }
        }
    }

    private String getPatchFilePath() {
        return SDCARD_PATH + PATCH_FILE;
    }

    private String getNewApkFilePath() {
        return SDCARD_PATH + NEW_APK_FILE;
    }

    /**
     * 打印日志
     * @param log
     */
    private void Log(String log) {
        Log.e("MainActivity", log);
    }

}

File: PatchUpdate/BsPatchKotlin/app/src/main/java/com/maotou/bspatch_kotlin/MainActivity.kt
package com.maotou.bspatch_kotlin

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.util.Log
import com.maotou.bspatch_kotlin.dialog.LoadDialog
import com.maotou.bspatch_kotlin.dialog.NormalDialog
import com.maotou.bspatch_kotlin.dialog.UpdateDialog
import com.maotou.bspatch_kotlin.httpservice.beans.UpdateBean
import com.maotou.bspatch_kotlin.httpservice.service.HttpMethod
import com.maotou.bspatch_kotlin.httpservice.serviceapi.UserApi
import com.maotou.bspatch_kotlin.httpservice.subscribers.HttpSubscriber
import com.maotou.bspatch_kotlin.httpservice.subscribers.SubscriberOnListener
import com.maotou.bspatch_kotlin.ui.UserServiceTestActivity
import com.maotou.bspatch_kotlin.util.ApkUtils
import com.maotou.bspatch_kotlin.util.CommonUtil
import com.maotou.bspatch_kotlin.util.SignUtils
import kotlinx.android.synthetic.main.activity_main.*
import org.jetbrains.anko.toast
import java.io.File


class MainActivity : AppCompatActivity() {

    var loadDialog: LoadDialog? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), 0)
        }

        bt_user_test.setOnClickListener {
            startActivity(Intent(this,UserServiceTestActivity::class.java))
        }

        bt_bspatch.setOnClickListener{
            patch()
        }
    }

    fun patch(){
        showDialog("检查更新")
        val path = ApkUtils.getCurApkPath(this)
        Log.d("patch-path",path)
        val file = File(path)
        val md5 = SignUtils.getMd5ByFile(file)
        Log.d("patch-md5",md5)
        UserApi.instance.checkUpdate(md5,ApkUtils.getVersionCode(this),"xiaomi",
                HttpSubscriber<UpdateBean>(object : SubscriberOnListener<UpdateBean>{
            override fun onSucceed(data: UpdateBean) {
                hideLoadDialog()

                val normalDialog = NormalDialog(this@MainActivity)
                normalDialog.show()
                val updatestr = "有新版本：${data.newVersionName}\n完整大小：${CommonUtil.getFormatSize(data.fileSize.toDouble())}\n增量大小：${CommonUtil.getFormatSize(data.patchSize.toDouble())}"
                normalDialog.setData("更新提醒",updatestr,"暂不更新","立即更新",object : NormalDialog.OnNoOrYesListener{
                    override fun onNo() {
                        //TODO 暂不更新
                    }

                    override fun onYes() {

                        if (ContextCompat.checkSelfPermission(this@MainActivity, Manifest.permission.WRITE_EXTERNAL_STORAGE)
                                != PackageManager.PERMISSION_GRANTED) {
                            toast("没有读写权限")
                            return
                        }

                        try {
                            val baseDir = "/maotou/"
                            val newApkPath = CommonUtil.makeDir(baseDir)
                            val newApkFile = File("$newApkPath/${packageName}_${data.newVersionName}.apk")
                            if (newApkFile.exists()) {
                                if (newApkFile.length() == data.fileSize) {
                                    ApkUtils.installApk(this@MainActivity, newApkFile.absolutePath)
                                    return
                                }
                            } else {
                                newApkFile.createNewFile()
                            }

                            val newPatchPath = CommonUtil.makeDir("${baseDir}patch")
                            val newPatchFile = File("${newPatchPath}/${packageName}_patch_${data.newVersionCode}_${data.versionCode}.patch")
                            if (!newPatchFile.exists()) {
                                newPatchFile.createNewFile()
                            }
                            val updateDialog = UpdateDialog(this@MainActivity)
                            updateDialog.show()
                            updateDialog.setInfo(HttpMethod.Download_URL + data.patchDownloadPath, newPatchFile.absolutePath, newApkFile.absolutePath, data.patchSize, data.fileSize)
                        }catch (e:Exception){
                            e.printStackTrace()
                        }
                    }

                })
            }

            override fun onError(code: Int, msg: String) {
                hideLoadDialog()
                toast(msg)
            }

        },this))
    }

    fun showDialog(msg: String) {
        if (loadDialog == null) {
            loadDialog = LoadDialog(this)
        }
        loadDialog?.show()
        loadDialog?.setLoadMsg(msg)
    }

    fun hideLoadDialog() {
        if (loadDialog !== null && loadDialog?.isShowing!!) {
            loadDialog?.dismiss()
        }
    }
}
